feat(config): migrate inline with the same sort

pret-a-porter opened the pull request:
## Changes:

Refactor migrations to provide ability to migrate and save original order of properties in configuration

## Context:

#11459

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

rarkins commented:
OK please get one step working first to demonstrate. I was hoping you would use the code I posted earlier which solves the majority of the problem in one go

HonkingGoose commented:
From our Pull Request template:

> Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. 
> 
> This makes it harder for us to review your work because we don't know what has changed. 
> 
> PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch.

rarkins commented:
@pret-a-porter could you summarize how much you've implemented, how it can be tested/verified, etc? e.g. I see one test which compares Object.keys(), is that it? It's the sorting of keys which is most important, so standard jest equivalency tests won't properly verify because they sort keys before comparing.

viceice commented on lib/config/migration.ts:
```diff
@@ -50,7 +33,8 @@ export function migrateConfig(
     }
-    const migratedConfig = clone(config) as MigratedRenovateConfig;
+    const newConfig = MigrationsService.run(config);
+    const migratedConfig = clone(newConfig) as MigratedRenovateConfig;
```
@rarkins shouldn't the clone be done in migration service?

rarkins replied:
Yes, I think so. @pret-a-porter any reason that won't work?

pret-a-porter replied:
To save order of properties, instead of clone original config we create empty object `const migratedConfig: RenovateConfig = {};` and then fill it step by step go over properties of original config

rarkins replied:
Does it need cloning? And if so, can the clone be done before the `return` statement instead of after?

pret-a-porter replied:
Yeah, we still need cloning. Maybe when all migrations will be refactored will be possible not to use `clone`.

We need cloning, because we use `dequal` to compare original and migrated configs. https://github.com/lukeed/dequal/blob/master/src/index.js under the hood this library by default compares values by `===` (it is not very clear from the docs).

I have tried to do `clone` before statement, but in this case some unit tests failed. I did not investigate deeper, if it is important I will do.

viceice replied:
Ok, so leave as is for now

viceice commented on lib/config/migrations/custom/go-mod-tidy-migration.spec.ts:
```diff
@@ -0,0 +1,32 @@
+    });
+
+    expect(migratedConfig.gomodTidy).toBeUndefined();
+    expect(migratedConfig.postUpdateOptions).toContain('test');
+    expect(migratedConfig.postUpdateOptions).toContain('gomodTidy');
```
Isn't a `toEqual([...])` simpler and cleaner  here, as used in other tests.

viceice commented on lib/config/migrations/base/abstract-migration.ts:
```diff
@@ -0,0 +1,26 @@
+
+  abstract run(): void;
+
+  protected delete(property: string): void {
```
What about making property optional and use `this.PropertyName` then?

This would make some migrations shorter and simpler. 🤗

pret-a-porter replied:
I thought about it as well, but this code `this.delete()` inside migration does not look descriptive for me

viceice commented on lib/config/migrations/migrations-service.spec.ts:
```diff
@@ -0,0 +1,45 @@
+      };
+
+      const migratedConfig = MigrationsService.run(originalConfig);
+      expect(migratedConfig[property]).toBeUndefined();
```
I think we should use `not.toHaveProperty` matcher?

Your code will not catch properties with `undefined` value assignment.

viceice commented on lib/config/migrations/migrations-service.spec.ts:
```diff
@@ -0,0 +1,45 @@
+      };
+
+      const migratedConfig = MigrationsService.run(originalConfig);
+      expect(migratedConfig[oldPropertyName]).toBeUndefined();
```
See comment above.

pret-a-porter commented:
> @pret-a-porter could you summarize how much you've implemented, how it can be tested/verified, etc? e.g. I see one test which compares Object.keys(), is that it? It's the sorting of keys which is most important, so standard jest equivalency tests won't properly verify because they sort keys before comparing.

To summarize:

1. Implemented class MigrationService, which should run migrations and save original order of properties. Basically yes, only one test checks that service save order of properties properly. I have tried to use Jest snapshots, but looks like it sorts properties before creating snapshot
2. Implemented basic migrations, to handle simple cases with removing and renaming properties in configuration
3. Implemented few custom migrations, to show how to handle more complicated migrations

rarkins commented:
@pret-a-porter can you provide some logs of comparing the migrated config log of this new code compared to the existing code? i.e. on "real" repos

rarkins commented:
The whole purpose of this issue was key ordering in migrated configs but there's almost no testing of that

pret-a-porter commented:
![image](https://user-images.githubusercontent.com/8514156/140625240-7ec39f93-56ff-4f1d-b2f0-b764882e8a9c.png)
@rarkins I have tried to run against https://github.com/pret-a-porter-testing/renovate-reproduction-11459/blob/main/renovate.json

rarkins commented:
> ![image](https://user-images.githubusercontent.com/8514156/140625240-7ec39f93-56ff-4f1d-b2f0-b764882e8a9c.png) @rarkins I have tried to run against [pret-a-porter-testing/renovate-reproduction-11459@`main`/renovate.json](https://github.com/pret-a-porter-testing/renovate-reproduction-11459/blob/main/renovate.json?rgh-link-date=2021-11-06T22%3A13%3A44Z)

When you test, it's best if you can include some *non* migrated config too, so that we can observe they are ordered correctly even if not migrated.

pret-a-porter commented:
@rarkins sorry, I did not get it. In have created repository with non migrated config https://github.com/pret-a-porter-testing/renovate-reproduction-11459. On screenshot you can see lines with `oldConfig` and `newConfig`. Shall I do something else?

rarkins commented:
> @rarkins sorry, I did not get it. I have created repository with non migrated config [pret-a-porter-testing/renovate-reproduction-11459](https://github.com/pret-a-porter-testing/renovate-reproduction-11459). On screenshot you can see lines with `oldConfig` and `newConfig`. Shall I do something else?

Yes, you need to do something else. Every config option in newConfig has been migrated. You should test with a mix of migrated and non-migrated (does not need migration) config.

pret-a-porter commented:
<img width="833" alt="Screenshot 2021-11-07 at 15 41 25" src="https://user-images.githubusercontent.com/8514156/140649696-8657eaab-b52b-4940-acae-7ff1e6a8a523.png">
@rarkins done

rarkins commented:
Can you put ignoreTests last and verify that it remains last?

rarkins commented:
Actually, have a mix of non-migrated fields first, last and in the middle of migrated and verify ordering remains as expected

pret-a-porter commented:
![image](https://user-images.githubusercontent.com/8514156/140653021-aee13e78-be5f-4f40-b3f6-e3cef57acfb3.png)
@rarkins done as well

pret-a-porter commented:
@rarkins shall I do any other validations regarding this PR?

rarkins commented:
> ![image](https://user-images.githubusercontent.com/8514156/140653021-aee13e78-be5f-4f40-b3f6-e3cef57acfb3.png) @rarkins done as well

This test doesn't interleave migrated and non-migrated. It has all migrated first, then non-migrated. Can you e.g. move `versionScheme` last in the old config?

pret-a-porter commented:
<img width="401" alt="Screenshot 2021-11-13 at 21 44 06" src="https://user-images.githubusercontent.com/8514156/141658578-e9ac1a32-8d78-478b-bbcb-2bcfa174f24a.png">
@rarkins it works fine

HonkingGoose commented on lib/config/migration.ts:
```diff
@@ -5,7 +5,7 @@ import { logger } from '../logger';
 import { clone } from '../util/clone';
```
Question: does this PR fully resolve issue #11459? Or is more work needed after this PR gets merged?

pret-a-porter replied:
Not fully. All migrations should be refactored by the new way https://github.com/renovatebot/renovate/issues/11457. Sorting of refactored migrated properties will be saved.
